# П.35. Отложить и выйти

**Defer** используется для завершения выполнения оператора непосредственно перед завершением функционального блока. Принимая во внимание, что **Exit** используется, чтобы заставить программу остановиться (помните, остановка программы, а не `return` просто остановка блока кода).

## A.35.1. Применение ключевых слов `defer`

Как кратко объяснено выше, это defer используется для завершения выполнения строки кода. Когда выполнение блока почти завершено, выполняется отложенный оператор.

Отсрочка может быть размещена где угодно, в начале или в конце блока. Но это не влияет на время выполнения, оно всегда будет выполняться в конце.

```
package main

import "fmt"

func main() {
    defer fmt.Println("halo")
    fmt.Println("selamat datang")
}

```

Выход:

![Применение `defer`](https://dasarpemrogramangolang.novalagung.com/images/A.35_1_defer.png)

Приведенное `defer` выше ключевое слово завершит исключение `fmt.Println("halo")` , эффект сообщения `"halo"` появится после `"selamat datang"` .

Отклоненный оператор все равно будет отображаться, даже если блок кода остановлен во время использования `return` . Например, как в следующем коде.

```
func main() {
    orderSomeFood("pizza")
    orderSomeFood("burger")
}

func orderSomeFood(menu string) {
    defer fmt.Println("Terimakasih, silakan tunggu")
    if menu == "pizza" {
        fmt.Print("Pilihan tepat!", " ")
        fmt.Print("Pizza ditempat kami paling enak!", "\n")
        return
    }

    fmt.Println("Pesanan anda:", menu)
}

```

Выход:

![Реализуйте `defer` с` return`](https://dasarpemrogramangolang.novalagung.com/images/A.35_2_defer_return.png)

Дополнительная информация, когда многие операторы отклоняются, все они будут выполняться в конце по порядку.

## A.35.2. Применение функций `os.Exit()`

Выход используется для одновременной принудительной остановки программы. Все операторы после выхода не будут выполнены, включая отложенные.

Функции `os.Exit()` в пакете `os` . Эта функция имеет числовой параметр типа, который должен быть заполнен. Введенный номер будет отображаться как **статус выхода** после остановки программы.

```
package main

import "fmt"
import "os"

func main() {
    defer fmt.Println("halo")
    os.Exit(1)
    fmt.Println("selamat datang")
}

```

Несмотря на то , что оператор `defer fmt.Println("halo")` размещен ранее `os.Exit()` , оператор не будет выполнен, потому что в середине функции программы принудительно останавливается.
